### PR TITLE
Fix optimize_skip_unused_shards for zero shards case

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix possible error ` Cannot find column` when `optimize_skip_unused_shards` is enabled and zero shards are used.

Detailed description / Documentation draft:
Before this patch it was using wrong header and cannot find column ignore() (for the query from the test):

```
Received from localhost:9000. DB::Exception: Cannot find column `ignore(1)` in source stream.
```

Cc: @amosbird 
Cc: @KochetovNicolai

### CODE REVIEW

File: src/Storages/StorageDistributed.cpp [hunk 1]
```diff
@@ -546,6 +549,14 @@ void StorageDistributed::read(
 
     ClusterProxy::executeQuery(query_plan, select_stream_factory, log,
         modified_query_ast, context, query_info);
+
+    if (!query_plan.isInitialized())
```
Comment (azat) [inline src/Storages/StorageDistributed.cpp:553]: Note that this place conflicts with #21550

Comment (amosbird) [reply src/Storages/StorageDistributed.cpp:553]: Are there any cases where query_plan is not initialized other than empty shard set? If no, I'd suggest moving the fix to the top of this function body.

Comment (azat) [reply src/Storages/StorageDistributed.cpp:553]: And in this case it may also conflicts with #21526 , but will see

**Review by KochetovNicolai — APPROVED** [review]

**Comment by azat** [comment]
>Testflows check — failed: 751, passed: 2688, other: 26

unrelated and broken

**Comment by azat** [comment]
>AST fuzzer (UBSan) — ../src/Interpreters/ExpressionActions.h:182:33: runtime error: member call on null pointer of type 'DB::ActionsDAG'

#21646

**Comment by KochetovNicolai** [comment]
Build is failed because of ci issue.

**Pull request merged**